File: 2016/18/18.py
def preprocessing(input_):
    return [0] + [c == '^' for c in input_] +  [0]

def solver(tiles):
    safe_tiles = tiles.count(0) - 2
    for row in range(1, 400_000):
        if row == 40: yield safe_tiles
        tiles       = [0] + [tiles[i] ^ tiles[i + 2] for i in range(len(tiles) - 2)] + [0]
        safe_tiles += tiles.count(0) - 2
    yield safe_tiles

File: 2016/18/18.rs
aoc2016::main!(); 

fn generator(input: &str) -> Vec<bool> {
    input
    .chars()
    .map(|c| c == '^')
    .collect_vec()
}

fn part_1(input: Vec<bool>) -> usize {
    solver(input, 40)
}

fn part_2(input: Vec<bool>) -> usize {
    solver(input, 400000)
}

fn solver(input: Vec<bool>, rows: usize) -> usize {
    input.iter().filter(|&b| !b).count() +
    (0..rows - 1)
        .scan(input.to_vec(), |acc, _| {
            *acc = [[false].to_vec(), acc.to_vec(), [false].to_vec()]
            .concat()
            .windows(3)
            .map(|w| w[0] ^ w[2])
            .collect_vec();
            Some(acc.clone())})
        .map(|s| s.iter().filter(|&&b| !b).count())
        .sum::<usize>()
}